runtime/interpreter/interpreter_expression.go

```diff
@@ -222,6 +248,53 @@ func (interpreter *Interpreter) memberExpressionGetterSetter(memberExpression *a
 	}
 }
 
+func shouldReturnReference(parent, member Value) bool {
+	_, parentIsReference := parent.(ReferenceValue)
+	return parentIsReference && isContainerValue(member)
+}
+
+func isContainerValue(value Value) bool {
+	switch value := value.(type) {
+	case *CompositeValue,
+		*SimpleCompositeValue,
+		*DictionaryValue,
+		*ArrayValue:
+		return true
+	case *SomeValue:
+		return isContainerValue(value.value)
+	default:
+		return false
+	}
+}
+
+// getReferenceValue Returns a reference to a given value.
+// Reference to an optional should return an optional reference.
+// This has to be done recursively for nested optionals.
+// e.g.1: Given type T, this method returns &T.
+// e.g.2: Given T?, this returns (&T)?
+func (interpreter *Interpreter) getReferenceValue(value Value, semaType sema.Type) Value {
+	optionalType, ok := semaType.(*sema.OptionalType)
+	if ok {
+		semaType = optionalType.Type
+
+		// Because the boxing happens further down the code, it is possible
+		// to have a concrete value (non-some) for a place where optional is expected.
+		// Therefore, always unwrap the type, but only unwrap if the value is `SomeValue`.
+		//
+		// However, checker guarantees that the wise-versa doesn't happen.
+		// i.e: There will never be a `SomeValue`, with type being a non-optional.
+
+		if optionalValue, ok := value.(*SomeValue); ok {
+			value = optionalValue.value
+		}
+		innerValue := interpreter.getReferenceValue(value, semaType)
+		return NewSomeValueNonCopying(interpreter, innerValue)
+	}
+
+	interpreter.maybeTrackReferencedResourceKindedValue(value)
+	return NewEphemeralReferenceValue(interpreter, false, value, semaType)
```

Comment:
I know this feature is theoretically not dependent on entitlements, but it might be worthwhile to base this PR on `feature/entitlements` to save yourself the merge conflict later. Here you would need to propagate the parent entitlements through any potential entitlement mapping.

Reply:
Good idea, actually entitlements are needed for the follow-up FLIPs anyway.

Rebased to a new feature branch `feature/mutability-restrictions` which is a fork from `feature/entitlements`